### ISSUE TITLE
Add grace period in case of APM Server errors or is offline

### ISSUE THREAD

**Issue by watson** [elastic/apm-agent-nodejs#684]

**Comment by ashuSvirus** [comment]
Did we find some resolution for this yet?

**Comment by watson** [comment]
Not yet. I started elastic/apm-nodejs-http-client#17 a while back, but never finished it because it had some issues. I looked at it again yesterday, but didn't complete it yet.

Are you experiencing any issues related to this, or is it just a "nice to have"?

**Comment by ashuSvirus** [comment]
yes. when the server ram is not that high, the server will crash after giving this error on stdout.

**Comment by watson** [comment]
I've tried to recreate that specific issue without any luck. Even if I set up a sever receiving several 1000 requests per second and no APM Server to receive the data, the memory footprint remained completely flat. I'm not ruling out that there's something wrong, but I just haven't been able to reproduce it.

The way the agent is supposed to work, is that it should only ever have a single outgoing TCP socket and HTTP request to the APM Server and no internal buffer. So if the APM Server is down, you should not see any increase in resources. At least in theory.

That being said, we should really implement the graceful back-off algorithm, but based on my early testing I unfortunately don't think that is what's causing your problems.

Could you share your agent config with me? Maybe I can spot something else that might cause the high memory usage.

**Comment by ashuSvirus** [comment]
@watson below is my config file for starting apm.
```
apm_client = require('elastic-apm-node').start({
        serviceName: 'name',
        serverUrl: 'url',
        serverTimeout: 30
       });
```

**Comment by vrochamaia** [comment]
I'm also facing the same issue when the APM server is not reachable / offline.
`
APM Server transport error (ETIMEDOUT): connect ETIMEDOUT XX.X.XX.XX:8200
[nodemon] app crashed - waiting for file changes before starting...
`

**Comment by watson** [comment]
@vrochamaia Could I get you to set `captureExceptions: false` in the agent config and run it again to see if you get a stack trace?

**Comment by vrochamaia** [comment]
Sorry, the problem is not with this library. It's actually with the `@elastic/elasticsearch` one. But it's kinda confusing. 

Without  `captureExceptions: false`:
`APM Server transport error (ETIMEDOUT): connect ETIMEDOUT XX.X.XX.XX:8200 [nodemon] app crashed - waiting for file changes before starting...`

With  `captureExceptions: false`:
```
const cloudUrls = Buffer.from(id.split(':')[1], 'base64').toString().split('$')
TypeError: Cannot read property 'split' of undefined
.
.
.
[nodemon] app crashed - waiting for file changes before starting...
```

Elasticsearch config:
```
const client = new Client({
    node: process.env.ELASTICSEARCH_HOST, 
    cloud: {
        user: process.env.ELASTICSEARCH_USER,
        password: process.env.ELASTICSEARCH_PASSWORD
    }
})
```
I guess that the problem is with that `cloud` option. The `ELASTICSEARCH_HOST` is accessible. If I remove the `cloud` option everything works fine.
